src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -101,3 +113,117 @@ func (r *downloadRequest) Validate() *util.JSONResponse {
+			}
+		}
+		// TODO: If we do not have a record and the origin is remote, we need to fetch it and respond with that file
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	// If we have a record, we can respond from the local file
+	r.MediaMetadata = mediaMetadata
+	return r.respondFromLocalFile(w, cfg.AbsBasePath)
+}
+
+// respondFromLocalFile reads a file from local storage and writes it to the http.ResponseWriter
+// Returns a util.JSONResponse error in case of error
+func (r *downloadRequest) respondFromLocalFile(w http.ResponseWriter, absBasePath types.Path) *util.JSONResponse {
+	filePath, err := fileutils.GetPathFromBase64Hash(r.MediaMetadata.Base64Hash, absBasePath)
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to get file path from metadata")
+		return &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.InternalServerError(),
+		}
+	}
+	file, err := os.Open(filePath)
+	defer file.Close()
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to open file")
+		return &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.InternalServerError(),
+		}
+	}
+	stat, err := file.Stat()
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to stat file")
+		return &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.InternalServerError(),
+		}
+	}
+
+	if r.MediaMetadata.FileSizeBytes > 0 && int64(r.MediaMetadata.FileSizeBytes) != stat.Size() {
+		r.Logger.WithFields(log.Fields{
+			"fileSizeDatabase": r.MediaMetadata.FileSizeBytes,
+			"fileSizeDisk":     stat.Size(),
+		}).Warn("File size in database and on-disk differ.")
+		return &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.InternalServerError(),
+		}
+	}
+
+	r.Logger.WithFields(log.Fields{
+		"UploadName":    r.MediaMetadata.UploadName,
+		"Base64Hash":    r.MediaMetadata.Base64Hash,
+		"FileSizeBytes": r.MediaMetadata.FileSizeBytes,
+		"Content-Type":  r.MediaMetadata.ContentType,
+	}).Info("Responding with file")
+
+	w.Header().Set("Content-Type", string(r.MediaMetadata.ContentType))
+	w.Header().Set("Content-Length", strconv.FormatInt(int64(r.MediaMetadata.FileSizeBytes), 10))
+	contentSecurityPolicy := "default-src 'none';" +
+		" script-src 'none';" +
+		" plugin-types application/pdf;" +
+		" style-src 'unsafe-inline';" +
+		" object-src 'self';"
+	w.Header().Set("Content-Security-Policy", contentSecurityPolicy)
+
+	if bytesResponded, err := io.Copy(w, file); err != nil {
+		r.Logger.WithError(err).Warn("Failed to copy from cache")
+		if bytesResponded == 0 {
+			return &util.JSONResponse{
+				Code: 500,
+				JSON: jsonerror.NotFound(fmt.Sprintf("Failed to respond with file with media ID %q", r.MediaMetadata.MediaID)),
+			}
+		}
+		// If we have written any data then we have already responded with 200 OK and all we can do is close the connection
```

Comment:
I'm going to explicitly `return nil` inside, at the end of the `if {}` block to avoid possible future bugs doing things after that block returns control to the remainder of the function. Done.